### DISCUSSION TITLE
Error handling/Catching mistakes when submitting something in admin_config

### DISCUSSION THREAD

**Discussion by LaocheXe** [e107inc/e107#5023] [Q&A Plugins]
Working on a plugin, and I am trying to add a "error message to admin/user when they submit something that is already in the database table, and if they left something blank.

I thought putting in an if statement in beforeCreate($new_data,$old_data) would work, but seems it's being ignored, and still submits the data (which in my case for testing, nothing is filled out.)

Or should I have put that in the onCreateError section?

**Comment by Jimmi08** [comment]
add required parameter to needed field
maybe you should use beforeUpdate() if it is editing. beforeCreate() is for the new record only.
BeforeCreate for the new record works, can you add the example of your code? 
And you still can use the type method and do all checks and messages yourself.

**Reply by LaocheXe** [reply]
For starters, I have an Dealership ID section that is required to select something (but it says Optional for some reason)

In the beforeCreate section I have:

                        if($new_data['ds_id'] == null)
			{
				// Error Message
				//print_a(LAN_DEALERSHIP_MISSING_ERROR);
				$errMsg = LAN_DEALERSHIP_MISSING_ERROR;
				return $errMsg;
			}

When I hit the submit button without selecting a dealership - it still goes through.

**Reply by Jimmi08** [reply]
use var_dump() not print_a() - otherwise you don't know if it is NULL or 0 or empty string. 
you need to return false or nothing I think if you want to stop inserting.  You should use e107::getMessage() handler to display the message.      

Write your code, it could ten reasons for "but it says Optional for some reason"

**Reply by LaocheXe** [reply]
var_dump() didn't display anything. Also the Optional thing is something else, so we don't need to worry about that. Here is the full code of beforeCreate:

public function beforeCreate($new_data, $old_data)
		{
			// Make sure the FiveM name isn't already there.
			$fivem_name = $new_data['v_fmname'];
			
			$sql = e107::getDb();
			
			$check_fivem_name = $sql->count('vehicles_exe', 'v_fmname', 'v_fmname = '.$fivem_name.'');
			
			var_dump($new_data['ds_id']);
			var_dump($new_data['vt_id']);
			var_dump($new_data['vb_id']);
			
			if($new_data['ds_id'] == null)
			{
				// Error Message
				//print_a(LAN_DEALERSHIP_MISSING_ERROR);
				$errMsg = LAN_DEALERSHIP_MISSING_ERROR;
				return $errMsg;
			}
			elseif($new_data['vt_id'] == 0)
			{
				$errMsg = "Please select a body type for the vehicle.";
				return $errMsg;
			}
			elseif($new_data['vb_id'] == 0)
			{
				$errMsg = "Please select the brand of the vehicle.";
				return $errMsg;
			}
			elseif($check_fivem_name > 0)
			{
				//print_a(LAN_FIVEM_NAME_ERROR);
				$errMsg = LAN_FIVEM_NAME_ERROR;
				return $errMsg;
			}
			else
			{
				return $new_data;
			}
			//return $new_data;
		}

**Reply by Jimmi08** [reply]
this `return $errMsg;` is true so it is continuing with inserting.   You can't return any text! 
var_dump() has to display something, try to combination with die(); before pushing submit button.

**Reply by LaocheXe** [reply]
I got it to send the message, but it won't stop the submission.

I removed the return $new_data but seems to do nothing.

	if(empty($new_data['ds_id']))
			{
				e107::getMessage()->addWarning(LAN_DEALERSHIP_MISSING_ERROR);
			}
			if(empty($new_data['vt_id']) || $new_data['vt_id'] == 0)
			{
				e107::getMessage()->addWarning('Please select a body type.');
			}
			if(empty($new_data['vb_id']) || $new_data['vb_id'] == 0)
			{
				e107::getMessage()->addWarning('Please select a brand.');
			}

**Reply by LaocheXe** [reply]
So to make the data required, I would add to the array 'validate' => 'value'   - however, can I change the 'value' to something else? There isn't anything in the documentation about it - when I add that to the dealership_id, it won't let me continue, even when I do select a dealership. Or do I have to setup the value some where else in the admin_config?

**Reply by Jimmi08** [reply]
I have no idea, I think that it is write parameter 'required'=> 1, or single parameter 'required'=> 1
there is test `if($data === false)`  - with 3x= it has to be boolean value returned so return false, not nothing.
I hope I will do some stuff with the new version next month so I will find some real examples. I don't have time now. You forget this stuff when you don't work with it too often.
Are you sure that you are using this for new records only?

**Reply by LaocheXe** [reply]
'validate' => true

**Reply by Jimmi08** [reply]
I don't know what validate is for, sorry, I always use required.

**Reply by LaocheXe** [reply]
That is requirement, makes the field required.

**Comment by Moc** [comment]
```
'validate'		=> true,
```

This indeed sets the field as required. 

@LaocheXe For checking if a field is filled in, use the validate option. Do you have other needed tests? Those can go in the public function `beforeCreate()` method (and `beforeUpdate()`)

You can try my Cookbook plugin, or Wrapper to look for inspiration.

**Reply by LaocheXe** [reply]
I guess that the count, check if one field is already in database wont work like it used to?

I have this:
$check_fivem_name = $sql->count('vehicles_exe', 'v_fmname', 'v_fmname = '.$fivem_name.'');
if($check_fivem_name > 0)
			{
				$errMsg = "error message here";
				return $errMsg;
			}
			else
			{
				return $new_data;
			}

so my question is, does the 'filter' do anything like this, would 'filter' be used for this?

**Reply by Moc** [reply]
Filter is used for filtering options like these: 

![image](https://github.com/e107inc/e107/assets/577978/44ae93e9-50b6-4315-af6a-7b47b9dba97e)


In your case, I would use this code: 

```php
public function beforeCreate($new_data, $old_data)
{

	$check_fivem_name = e107::getDb()->count('vehicles_exe', 'v_fmname', 'v_fmname = '.$new_data['fivem_name'].'');
	
	if($check_fivem_name > 0)
	{
		e107::getMessage()->addError("error message here");
		return false; 
	}
	
	return $new_data;
}
```
